infrastructures/infrastructure-openstack/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/OpenstackInfrastructure.java

```diff
@@ -325,4 +357,241 @@ protected void unregisterNodeAndRemoveInstanceIfNeeded(final String instanceTag,
             return null;
         });
     }
+
+    private void internalAcquireNodes(int numberOfNodes, Map<String, ?> dynamicPolicyParameters) {
+
+        nbOfNodes = numberOfNodes;
+
+        if (nbOfNodes > 0 && checkDynamicPolicyParameters(nbOfNodes, dynamicPolicyParameters)) {
+
+            // Create Openstack infrastructure (if it does not exist) and initialize its persistent variables
+            initializeOpenstackInfrastructure();
```

Comment:
What if infrastructure already exists. How do we recover VMs? Also how do we recover deploying VMs? 
I am not saying something is wrong, I just want to be sure you have answers)

Reply:
If the infrastructure already exists, the init step is skipped, only instances (VMs) and nodes are added to that infra. 
At the deployment phase, instances (VMs) and nodes have two statuses. Either the deployment succeeds, or they are removed (zombie VMs and down/lost nodes are removed).